builtins/amp-img.js

```diff
@@ -132,10 +145,67 @@ export class AmpImg extends BaseElement {
 
     this.propagateAttributes(ATTRIBUTES_TO_PROPAGATE, this.img_);
     guaranteeSrcForSrcsetUnsupportedBrowsers(this.img_);
+    if (isExperimentOn(this.getAmpDoc().win, 'amp-img-auto-sizes')) {
+      this.maybeGenerateSizes_();
+    }
     this.applyFillContent(this.img_, true);
+
     this.element.appendChild(this.img_);
   }
 
+  /**
+   * This function automatically generates sizes for amp-imgs without
+   * the sizes attribute.
+   * @private
+   */
+  maybeGenerateSizes_() {
+    if (!this.img_) {
+      return;
+    }
+    // No need to generate sizes if already present.
+    const sizes = this.element.getAttribute('sizes');
+    if (sizes) {
+      return;
+    }
+    // Sizes is useless without the srcset attribute or if the srcset
+    // attribute uses the x descriptor.
+    const srcset = this.element.getAttribute('srcset');
+    if (!srcset || /[0-9]+x(?:,|$)/.test(srcset)) {
+      return;
+    }
+
+    const width = this.getLayoutWidth();
+    const viewportWidth = this.getViewport().getWidth();
```

Comment:
We can check `shouldSetSizes` here and early exit.

Reply:
Done.